File: khmer-wasm/tests/segmenter.test.ts
/**
 * Unit tests for Khmer Word Segmenter (WASM/AssemblyScript).
 * Tests against the shared test cases to ensure 100% match with Python baseline.
 */

import { describe, it, expect, beforeAll } from 'vitest';
import * as fs from 'fs';
import * as path from 'path';

// WASM module exports
let wasmModule: any;
let segmenter: any;

interface TestCase {
    id: number;
    input: string;
    description: string;
    expected: string[];
}

describe('KhmerSegmenter WASM', () => {
    let testCases: TestCase[];

    beforeAll(async () => {
        // Find data directory
        let dataDir = path.join(__dirname, '..', '..', 'data');
        if (!fs.existsSync(dataDir)) {
            dataDir = path.join(__dirname, '..', 'data');
        }

        const dictPath = path.join(dataDir, 'khmer_dictionary_words.txt');
        const freqPath = path.join(dataDir, 'khmer_word_frequencies.json');
        const testCasesPath = path.join(dataDir, 'test_cases.json');

        // Load WASM module
        const wasmPath = path.join(__dirname, '..', 'build', 'release.js');
        wasmModule = await import(wasmPath);

        // Initialize dictionary and segmenter
        const dictContent = fs.readFileSync(dictPath, 'utf-8');
        const freqContent = fs.readFileSync(freqPath, 'utf-8');

        wasmModule.initDictionary(dictContent, freqContent);

        // Load test cases
        testCases = JSON.parse(fs.readFileSync(testCasesPath, 'utf-8'));
    });

    it('should match all expected outputs', () => {
        const failures: string[] = [];

        for (const tc of testCases) {
            const resultJson = wasmModule.segment(tc.input);
            const result = JSON.parse(resultJson);

            if (JSON.stringify(result) !== JSON.stringify(tc.expected)) {
                failures.push(
                    `[${tc.id}] ${tc.description}\n` +
                    `  Input: ${tc.input}\n` +
                    `  Expected: ${JSON.stringify(tc.expected)}\n` +
                    `  Actual: ${JSON.stringify(result)}`
                );
            }
        }

        if (failures.length > 0) {
            throw new Error(
                `${failures.length}/${testCases.length} test cases failed:\n${failures.join('\n')}`
            );
        }
    });

    it('should segment single known word', () => {
        const result = JSON.parse(wasmModule.segment('សួស្តី'));
        expect(result).toEqual(['សួស្តី']);
    });

    it('should segment multiple words', () => {
        const result = JSON.parse(wasmModule.segment('ខ្ញុំស្រលាញ់កម្ពុជា'));
        expect(result).toEqual(['ខ្ញុំ', 'ស្រលាញ់', 'កម្ពុជា']);
    });

    it('should preserve spaces', () => {
        const result = JSON.parse(wasmModule.segment('សួស្តី បង'));
        expect(result).toEqual(['សួស្តី', ' ', 'បង']);
    });

    it('should handle numbers', () => {
        const result = JSON.parse(wasmModule.segment('១២៣៤៥'));
        expect(result).toEqual(['១២៣៤៥']);
    });

    it('should handle empty string', () => {
        const result = JSON.parse(wasmModule.segment(''));
        expect(result).toEqual([]);
    });

    it('should handle space before sign pattern (regression)', () => {
        const result = JSON.parse(wasmModule.segment('សម្រា ប់ការ'));
        expect(result).toEqual(['ស', 'ម្រា ប់', 'ការ']);
    });

    it('should handle punctuation', () => {
        const result = JSON.parse(wasmModule.segment('សួស្តី។'));
        expect(result).toEqual(['សួស្តី', '។']);
    });
});

File: khmer-go/pkg/khmer/segmenter_test.go
package khmer

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

// TestCase represents a single test case from the shared test file
type TestCase struct {
	ID          int      `json:"id"`
	Input       string   `json:"input"`
	Description string   `json:"description"`
	Expected    []string `json:"expected"`
}

var testSegmenter *KhmerSegmenter
var testCases []TestCase

func TestMain(m *testing.M) {
	// Find data directory (try multiple locations)
	possiblePaths := []string{
		"../../data",           // From pkg/khmer/
		"../../../data",        // From pkg/khmer/ with extra level
		"data",                 // Current directory
		"../data",              // Parent directory
		"../../../../data",     // Deep nesting
	}

	dataDir := ""
	for _, p := range possiblePaths {
		testFile := filepath.Join(p, "test_cases.json")
		if _, err := os.Stat(testFile); err == nil {
			dataDir = p
			break
		}
	}
	if dataDir == "" {
		panic("Could not find data directory with test_cases.json")
	}

	dictPath := filepath.Join(dataDir, "khmer_dictionary_words.txt")
	freqPath := filepath.Join(dataDir, "khmer_word_frequencies.json")
	testCasesPath := filepath.Join(dataDir, "test_cases.json")

	// Initialize segmenter
	dict := NewDictionary()
	if err := dict.Load(dictPath, freqPath); err != nil {
		panic("Failed to load dictionary: " + err.Error())
	}
	testSegmenter = NewKhmerSegmenter(dict)

	// Load test cases
	data, err := os.ReadFile(testCasesPath)
	if err != nil {
		panic("Failed to load test cases: " + err.Error())
	}
	if err := json.Unmarshal(data, &testCases); err != nil {
		panic("Failed to parse test cases: " + err.Error())
	}

	os.Exit(m.Run())
}

func TestAllCasesMatchExpected(t *testing.T) {
	var failures []TestCase
	for _, tc := range testCases {
		result := testSegmenter.Segment(tc.Input)
		if !reflect.DeepEqual(result, tc.Expected) {
			failures = append(failures, TestCase{
				ID:          tc.ID,
				Input:       tc.Input,
				Description: tc.Description,
				Expected:    tc.Expected,
			})
			t.Errorf("[%d] %s\n  Input: %s\n  Expected: %v\n  Actual: %v",
				tc.ID, tc.Description, tc.Input, tc.Expected, result)
		}
	}
	if len(failures) > 0 {
		t.Errorf("%d/%d test cases failed", len(failures), len(testCases))
	}
}

func TestSingleKnownWord(t *testing.T) {
	result := testSegmenter.Segment("សួស្តី")
	expected := []string{"សួស្តី"}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("Expected %v, got %v", expected, result)
	}
}

func TestMultipleWords(t *testing.T) {
	result := testSegmenter.Segment("ខ្ញុំស្រលាញ់កម្ពុជា")
	expected := []string{"ខ្ញុំ", "ស្រលាញ់", "កម្ពុជា"}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("Expected %v, got %v", expected, result)
	}
}

func TestWithSpaces(t *testing.T) {
	result := testSegmenter.Segment("សួស្តី បង")
	expected := []string{"សួស្តី", " ", "បង"}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("Expected %v, got %v", expected, result)
	}
}

func TestNumbers(t *testing.T) {
	result := testSegmenter.Segment("១២៣៤៥")
	expected := []string{"១២៣៤៥"}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("Expected %v, got %v", expected, result)
	}
}

func TestEmptyString(t *testing.T) {
	result := testSegmenter.Segment("")
	if len(result) != 0 {
		t.Errorf("Expected empty slice, got %v", result)
	}
}

func TestSpaceBeforeSignPattern(t *testing.T) {
	// Regression test for the fix
	result := testSegmenter.Segment("សម្រា ប់ការ")
	expected := []string{"ស", "ម្រា ប់", "ការ"}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("Expected %v, got %v", expected, result)
	}
}

func TestPunctuation(t *testing.T) {
	result := testSegmenter.Segment("សួស្តី។")
	expected := []string{"សួស្តី", "។"}
	if !reflect.DeepEqual(result, expected) {
		t.Errorf("Expected %v, got %v", expected, result)
	}
}
